includes/Util/UrlResolver.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+declare(strict_types=1);
+
+namespace LabkiPackManager\Util;
+
+use MediaWiki\MediaWikiServices;
+
+/**
+ * UrlResolver
+ *
+ * Utility class for resolving repository URLs to git clone URLs.
+ * Handles GitHub URLs and converts them to appropriate git clone URLs.
+ */
+final class UrlResolver {
+
+    /**
+     * Resolve a repository URL to a git clone URL.
+     *
+     * Supported forms:
+     *  - github.com repo URLs (optionally with /tree/<ref> or /blob/<ref>/<path>)
+     *    are converted to git clone URLs with appropriate refs
+     *  - Already valid git clone URLs are returned as-is
+     *  - Other HTTP(S) URLs are returned as-is for potential git clone operations
+     *
+     * @param string $repoUrl
+     * @return string Git clone URL
+     */
+    public static function resolveContentRepoUrl(string $repoUrl): string {
+        $trimRepoUrl = trim($repoUrl);
+
+        if ($trimRepoUrl === '') {
+            return $trimRepoUrl;
+        }
+
+        // Parse URL components
+        $parts = @parse_url($trimRepoUrl) ?: [];
+        $originalHost = $parts['host'] ?? '';
+        $host = strtolower($originalHost);
+        $path = $parts['path'] ?? '';
+
+        // Helper to rebuild URL from parts
+        $rebuild = function(array $p, array $originalParts = []) use ($originalHost): string {
```

Comment:
incredible that php provides `parse_url` and not the inverse. this is sorta odd as a partial-merge-partial-rebuild function but it looks like it should work in this case at least. might run into trouble with how the merge is only partial and `scheme`, `host`, and `path` are all just empty strings if `p` doesnt' have them but updated from `originalParts` from the other parts, (using `ssh://` is somewhat common), but should be fine for github urls